src/cargo/sources/git/oxide.rs

```diff
@@ -49,22 +49,78 @@ pub fn with_retry_and_progress(
         .ok()
         .unwrap_or_default()
         .auto_deregister();
-    let should_interrupt = AtomicBool::new(false);
     let mut progress_bar = Progress::new("Fetch", config);
     std::thread::scope(move |s| {
         s.spawn({
             let root = Arc::downgrade(&progress_root);
             move || -> CargoResult<()> {
+                const READ_PACK_BYTES: [u8; 4] = *b"BWRB";
+                const DELTA_INDEX_OBJECTS: [u8; 4] = *b"IWIO";
+                const RESOLVE_OBJECTS: [u8; 4] = *b"IWRO";
+
+                // We choose `N=10` here to make a `300ms * 10slots ~= 3000ms`
+                // sliding window for tracking the data transfer rate (in bytes/s).
+                let mut last_update = Instant::now();
+                let mut counter = MetricsCounter::<10>::new(0, last_update);
+
                 let mut tasks = Vec::with_capacity(10);
                 while let Some(root) = root.upgrade() {
                     root.sorted_snapshot(&mut tasks);
-                    progress_bar.tick(0, 10, "TBD")?;
-                    // dbg!(&tasks);
+                    let counters = tasks
+                        .iter()
+                        .find_map(|(_, t)| {
+                            (t.id == READ_PACK_BYTES).then(|| t.progress.as_ref().expect("set"))
+                        })
+                        .and_then(|read| {
+                            tasks.iter().find_map(|(_, t)| {
+                                (t.id == DELTA_INDEX_OBJECTS)
+                                    .then(|| (t.progress.as_ref().expect("set"), Some(read)))
+                            })
+                        })
+                        .or_else(|| {
+                            tasks.iter().find_map(|(_, t)| {
+                                (t.id == RESOLVE_OBJECTS)
+                                    .then(|| (t.progress.as_ref().expect("set"), None))
+                            })
+                        });
+                    let progress_info = match counters {
```

Comment:
It's indeed something specific to `gitoxide` and thus necessary to get this to work in the simplest possible way.

More specifically, `prodash` implements a tree of counters. These keep information about the progress, but also plenty of meta-data to identify and display it (useful for things like units, and names). `prodash` provides renderers for these progress-trees which leverage that they are fully specified by themselves. All this is great if one wants to use one of these renderers and if one is happy with having very informative (but conmplex) progress information.

Having custom progress rendering means we have to fetch information out of this tree selectively and deal with the potential absence of one of many progress bars that we might be interested in.

Since `gitoxide` uses the `prodash::Progress` trait, it's possible to revert to a callback based mechanism by implementing it, and hook up the `cargo` progress bar directly. That way the system would be callback based like before and the progress bar _probably_ wouldn't have to be thread-safe. This would come at the cost of significantly more boilerplate, and in the end one would still have to navigate a tree of progress bars (this time within our own data structure), while being hammered by hundreds of thousands of progress calls per second as `gitoxide` assumes progress calls are as cheap as incrementing an atomic.